nuxeo-runtime/nuxeo-runtime-stream/src/main/java/org/nuxeo/runtime/stream/StreamServiceImpl.java

```diff
@@ -0,0 +1,214 @@
+            }
+            LogConfigDescriptor config = configs.get(name);
+            if (config.isKafkaLog()) {
+                managers.put(name, createKafkaLogManager(config));
+            } else {
+                managers.put(name, createChronicleLogManager(config));
+            }
+        }
+        return managers.get(name);
+    }
+
+    protected LogManager createKafkaLogManager(LogConfigDescriptor config) {
+        String kafkaConfig = config.getOption("kafkaConfig", "default");
+        KafkaConfigService service = Framework.getService(KafkaConfigService.class);
+        return new KafkaLogManager(service.getZkServers(kafkaConfig),
+                service.getTopicPrefix(kafkaConfig),
+                service.getProducerProperties(kafkaConfig),
+                service.getConsumerProperties(kafkaConfig));
+    }
+
+    protected LogManager createChronicleLogManager(LogConfigDescriptor config) {
+        String basePath = config.getOption("basePath", null);
+        String directory = config.getOption("directory", config.getName());
+        Path path = getChroniclePath(basePath, directory);
+        String retention = getChronicleRetention(config.getOption("retention", null));
+        return new ChronicleLogManager(path, retention);
+    }
+
+    protected String getChronicleRetention(String retention) {
+        return retention != null ? retention : Framework.getProperty(NUXEO_STREAM_RET_DURATION_PROP, "4d");
+    }
+
+    protected Path getChroniclePath(String basePath, String name) {
+        if (basePath != null) {
+            return Paths.get(basePath, name).toAbsolutePath();
+        }
+        basePath = Framework.getProperty(NUXEO_STREAM_DIR_PROP);
+        if (basePath != null) {
+            return Paths.get(basePath, name).toAbsolutePath();
+        }
+        basePath = Framework.getProperty(Environment.NUXEO_DATA_DIR);
+        if (basePath != null) {
+            return Paths.get(basePath, "stream", name).toAbsolutePath();
+        }
+        return Paths.get(Framework.getRuntime().getHome().getAbsolutePath(),
+                "data", "stream", name).toAbsolutePath();
+    }
+
+    protected void createStreamIfNotExists(String name, LogConfigDescriptor config) {
+        if (config.getLogsToCreate().isEmpty()) {
+            return;
+        }
+        LogManager manager = getLogManager(name);
+        config.getLogsToCreate()
+                .forEach((stream, size) -> {
+                    log.info("Create if not exists stream: " + stream + " with manager: " + name);
+                    manager.createIfNotExists(stream, size);
+                });
+    }
+
+    @Override
+    public void start(ComponentContext context) {
+        super.start(context);
+        configs.forEach(this::createStreamIfNotExists);
+        processorDescriptors.forEach(this::initProcessor);
+        Framework.getRuntime().getComponentManager().addListener(new ComponentsLifeCycleListener());
```

Comment:
ok I can add check so that addListener is done only once, do we have example of correct start impl with component listener ? the workmanagerImpl looks weird

Reply:
Not really :/ we should provide a better install/uninstall way for these listeners. The only way we have currently is to create it during activate/start steps and uninstall it during deactivate/stop steps.

You can also create it in the constructor, if you don't need it to be re-created, but I'm not sure if it's working.